test_util/test_upgrade_vpc.py

```diff
@@ -82,39 +119,94 @@ def get_test_app():
                 "path": "/",
                 "portIndex": 0,
                 "gracePeriodSeconds": 5,
-                "intervalSeconds": 10,
-                "timeoutSeconds": 10,
-                "maxConsecutiveFailures": 3
+                "intervalSeconds": 1,
+                "timeoutSeconds": 5,
+                "maxConsecutiveFailures": 1
             }
         ],
     }
-    return app
 
+    # DNS resolution app to make sure DNS is available during the upgrade.
+    # Periodically resolves the healthcheck app's domain name and logs whether
+    # it succeeded to a file in the Mesos sandbox.
+    dns_app = {
+        "id": '/' + TEST_APP_NAME_FMT.format('dns-' + test_id),
+        "cmd": """
+while true
+do
+    printf "%s " $(date --utc -Iseconds) >> $MESOS_SANDBOX/$DNS_LOG_FILENAME
+    if host -W $TIMEOUT_SECONDS $RESOLVE_NAME
+    then
+        echo SUCCESS >> $MESOS_SANDBOX/$DNS_LOG_FILENAME
+    else
+        echo FAILURE >> $MESOS_SANDBOX/$DNS_LOG_FILENAME
+    fi
+    sleep $INTERVAL_SECONDS
+done
+""",
+        "env": {
+            'RESOLVE_NAME': marathon_app_id_to_mesos_dns_name(healthcheck_app['id']) + '.marathon.mesos',
+            'DNS_LOG_FILENAME': 'dns_resolve_log.txt',
+            'INTERVAL_SECONDS': '1',
+            'TIMEOUT_SECONDS': '1',
+        },
+        "cpus": 0.5,
+        "mem": 32.0,
+        "instances": 1,
+        "container": {
+            "type": "DOCKER",
+            "docker": {
+                "image": "mapitman/bind-utils",
+                "network": "BRIDGE",
+            }
+        },
+        "dependencies": [healthcheck_app['id']],
+    }
 
-def get_task_info(apps, tasks):
-    idx = 0    # We have a single app and a task for this test.
+    # Deploy test apps.
+    # TODO(branden): Define apps and deploy_app() such that we can deploy all
+    # test apps at once and then block on waiting for them all to finish
+    # deploying. Marathon ought to be responsible for deploying apps in the
+    # right order and knowing whether they're deployed and stable.
+    cluster_api.marathon.deploy_app(healthcheck_app)
+    cluster_api.marathon.ensure_deployments_complete()
+    # TODO(branden): This is a hack to make sure we don't deploy dns_app before
+    # the name it's trying to resolve is available.
+    _wait_for_dns(dns_app['env']['RESOLVE_NAME'])
+    cluster_api.marathon.deploy_app(dns_app, check_health=False)
+    cluster_api.marathon.ensure_deployments_complete()
+
+    test_apps = [healthcheck_app, dns_app]
+    test_app_ids = [app['id'] for app in test_apps]
+
+    tasks_start = {app_id: _app_tasks(app_id) for app_id in test_app_ids}
```

Comment:
Not sure what value is added with an additional check that `/v2/apps` isn't empty. Either error suggests that Marathon lost track of test apps, and I'd go about debugging them in the same way (comparing the services UI against the Mesos UI, inspecting Marathon logs, etc.). It also unnecessarily locks us into only using this context manager on a cluster that has no running Marathon apps. We might want to run other test apps outside this function, or some day we may start running core DC/OS services as Marathon apps (once we get Mesos agents running on the masters and Marathon adds support for daemon tasks).

We're only looking for tasks for the specific test apps that we're tracking, so `/v2/apps/<app_id>/tasks` should be fine. I could use `/v2/tasks` here, but then I'd have to filter out unrelated tasks and group the rest by app, which is the convenience that `/v2/apps/<app_id>/tasks` is supposed to provide.